compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/CompilationRequestManager.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.ballerinalang.compiler;
+
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.StaleCompilationRequestException;
+
+/**
+ * This class is responsible for collecting compilation requests that are coming in from various sources and ensuring
+ * that the latest request is maintained.
+ *
+ * @since 2.0.0
+ */
+public class CompilationRequestManager {
+
+    private static final CompilerContext.Key<CompilationRequestManager> COMPILATION_REQUEST_MANAGER_KEY =
+            new CompilerContext.Key<>();
+
+    private CompilationRequest currentCompileRequest;
+
+    private CompilationRequestManager(CompilerContext context) {
+        context.put(COMPILATION_REQUEST_MANAGER_KEY, this);
+    }
+
+    public static synchronized CompilationRequestManager getInstance(CompilerContext context) {
+        CompilationRequestManager compilationRequestManager = context.get(COMPILATION_REQUEST_MANAGER_KEY);
+        if (compilationRequestManager == null) {
+            compilationRequestManager = new CompilationRequestManager(context);
+        }
+        return compilationRequestManager;
+    }
+
+    /**
+     * This method accepts a {@link CompilationRequest} instance and compares it against the currently held compilation
+     * request. If this new request is more recent than the currently held request, it will discard the current request
+     * and set this new request as the current request. The provided request must not be null.
+     *
+     * @param request A request to perform a compilation
+     */
+    public synchronized void submit(CompilationRequest request) {
+        if (currentCompileRequest == null || currentCompileRequest.getTimestamp().isBefore(request.getTimestamp())) {
```

Comment:
Do we really need to track the time here? Creating and comparing timestamp will slower down the comparison IMO. An `int` based **incrementing** **counter** would be ideal in this scenario. So you can compare freshness with primitive comparisons such as `>`.

Reply:
Should be fine I think. It's basically just a `long` comparison

Reply:
No, when you calling `Instant.now()` it causes a **sys call** to get current instant from the system UTC clock. Why we are unnecessarily keeping a timestamp?